pkg/apis/flowcontrol/v1beta3/conversion.go

```diff
@@ -0,0 +1,104 @@
+
+import (
+	"fmt"
+
+	"k8s.io/api/flowcontrol/v1beta3"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+)
+
+func Convert_v1beta3_PriorityLevelConfiguration_To_flowcontrol_PriorityLevelConfiguration(in *v1beta3.PriorityLevelConfiguration, out *flowcontrol.PriorityLevelConfiguration, s conversion.Scope) error {
+	if err := autoConvert_v1beta3_PriorityLevelConfiguration_To_flowcontrol_PriorityLevelConfiguration(in, out, nil); err != nil {
+		return err
+	}
+
+	// during v1beta3 -> internal conversion:
+	//  - remove the roundtrip annotation for the 'NominalConcurrencyShares' field
+	//  - make sure we don't mutate the source (v1beta3) object's annotations
+	annotations, copied := dropPriorityLevelConcurrencyShareDefaultAnnotation(in.ObjectMeta.Annotations)
+	if copied {
+		out.ObjectMeta.Annotations = annotations
+	}
+	return nil
+}
+
+func Convert_flowcontrol_PriorityLevelConfiguration_To_v1beta3_PriorityLevelConfiguration(in *flowcontrol.PriorityLevelConfiguration, out *v1beta3.PriorityLevelConfiguration, s conversion.Scope) error {
+	if err := autoConvert_flowcontrol_PriorityLevelConfiguration_To_v1beta3_PriorityLevelConfiguration(in, out, nil); err != nil {
+		return err
+	}
+
+	// during internal -> v1beta3 conversion:
+	//  - add the roundtrip annotation for the 'NominalConcurrencyShares' field,
+	//      IIF the 'NominalConcurrencyShares' field has a value of zero.
+	//  - make sure we don't mutate the source (internal) object's annotations
+	limited := in.Spec.Limited
+	if limited == nil {
+		return nil
+	}
+	if limited.NominalConcurrencyShares == nil {
+		// If not specified by the user, the defaulting logic is
+		// expected to set the 'NominalConcurrencyShares' field of a
+		// LimitedPriorityLevelConfiguration object to a default
+		// value, so after conversion, we don't expect the
+		// 'NominalConcurrencyShares' field in an
+		// internal object to be nil.
+		return fmt.Errorf("unexpected nil value for NominalConcurrencyShares in the internal object")
+	}
+	if *limited.NominalConcurrencyShares != 0 {
+		return nil
+	}
+
+	annotations, copied := addPriorityLevelConcurrencyShareDefaultAnnotation(in.ObjectMeta.Annotations)
+	if copied {
+		out.ObjectMeta.Annotations = annotations
+	}
+
+	return nil
+}
+
+func dropPriorityLevelConcurrencyShareDefaultAnnotation(in map[string]string) (map[string]string, bool) {
+	if _, ok := in[v1beta3.PriorityLevelConcurrencyShareDefaultKey]; !ok {
+		return in, false
+	}
+
+	out := deepCopyStringMap(in)
+	delete(out, v1beta3.PriorityLevelConcurrencyShareDefaultKey)
+	return out, true
+}
+
+func addPriorityLevelConcurrencyShareDefaultAnnotation(in map[string]string) (map[string]string, bool) {
+	if v, ok := in[v1beta3.PriorityLevelConcurrencyShareDefaultKey]; ok && v == "" {
+		return in, false
+	}
+
+	out := deepCopyStringMap(in)
+	out[v1beta3.PriorityLevelConcurrencyShareDefaultKey] = ""
+	return out, true
+}
+
+// deepCopyStringMap returns a copy of the input map.
```

Comment:
The word "deep" here is at best meaningless, at worst confusing.
There is only one level of mutable structure here, so there is no difference between "deep" and "shallow" copy.
Putting the word "deep" in the name suggests that there _is_ a difference.